checker/src/tag_domain.rs

```diff
@@ -8,92 +8,48 @@ use crate::bool_domain::BoolDomain;
-#[derive(Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
-pub struct SerializableDefId {
-    pub krate: CrateNum,
-    pub index: DefIndex,
-}
-
-impl From<DefId> for SerializableDefId {
-    fn from(did: DefId) -> SerializableDefId {
-        SerializableDefId {
-            krate: did.krate,
-            index: did.index,
-        }
-    }
-}
-
-impl std::fmt::Debug for SerializableDefId {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_fmt(format_args!(
-            "{:?}",
-            DefId {
-                krate: self.krate,
-                index: self.index
-            }
-        ))
-    }
+use serde::{Deserialize, Serialize};
+use std::rc::*;
+
+/// Tags are either a pre-defined tag for constant-time security, or user-defined ones.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Clone, Serialize, Deserialize, Hash)]
+pub enum Tag {
+    /// A tag for analyzing constant-time security. The type is pre-defined in mirai_annotations.
+    ConstantTime,
+
+    /// A user-defined tag is represented as a pair of its tag kind and its propagation set.
+    /// The tag kind is the path of the declared tag type, and the propagation set is associated to the
+    /// tag type as a const generic parameter.
+    Custom {
+        def_path_str: Rc<String>,
+        prop_set: TagPropagationSet,
+    },
 }
 
-impl Serialize for SerializableDefId {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let krate = u64::from(self.krate.as_u32());
-        let index = u64::from(self.index.as_u32());
-        serializer.serialize_u64((krate << 32) | index)
+impl Tag {
+    /// Check if a value of enum type `TagPropagation` is included in `self`'s propagation set.
+    #[logfn_inputs(TRACE)]
+    pub fn is_propagated_by(&self, exp_tag_prop: TagPropagation) -> bool {
+        precondition!((exp_tag_prop as u8) < 128);
+        self.get_prop_set() & (1 << (exp_tag_prop as u8)) != 0
     }
-}
 
-impl<'de> Deserialize<'de> for SerializableDefId {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        struct U64Visitor;
-
-        impl<'de> de::Visitor<'de> for U64Visitor {
-            type Value = u64;
-
-            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-                formatter.write_str("u64")
-            }
-
-            fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
-            where
-                E: de::Error,
-            {
-                Ok(value)
-            }
+    /// Obtain `self`'s propagation set. The constant-time tag will be propagated by all operations.
+    #[logfn_inputs(TRACE)]
+    pub fn get_prop_set(&self) -> TagPropagationSet {
+        match self {
+            Tag::ConstantTime => TAG_PROPAGATION_ALL,
```

Comment:
I'm not sure this is desirable. Some operations may randomize secrets enough to amount to sanitization. This design is not flexible.